src/update/update.cpp

```diff
@@ -0,0 +1,517 @@
+               result += VoteResult(yesCount, samples);
+            }
+
+            // LogPrint(BCLog::UPDATES, "Updates: NodeRoundVoting, Height: %d, Tx: %d, Quorum Commitment\n", curIndex->nHeight, i);
+         }
+      }
+   }
+   LogPrint(BCLog::UPDATES, "Updates: NodeRoundVoting::GetVote, Height: %7d, %s\n", blockIndex->nHeight, result.ToString().c_str());
+   cache[{ update.UpdateId(), blockIndex }] = result;
+   return result;
+}
+
+VoteResult MinerUpdateVoting::GetVote(const CBlockIndex* blockIndex, const Update &update)
+{
+   // assert(height % update.roundSize == 0);
+   if (blockIndex->nHeight < update.StartHeight() + update.RoundSize() * update.VotingPeriod())
+   {
+      return VoteResult(0, update.RoundSize() * update.VotingPeriod()); // Assume everyone voted no for the entire period
+   }
+   // TODO: Cache lookup for Update Height
+
+   // Simply collect all of the round votes for a full average in the voting period:
+   VoteResult vote;
+   for (int lookback = 0; lookback < update.VotingPeriod(); ++lookback)
+   {
+      const CBlockIndex* roundBlockIndex = blockIndex->GetAncestor(blockIndex->nHeight - lookback * update.RoundSize());
+      if (roundBlockIndex == nullptr)
+         break; // Nothing more
+      VoteResult roundVote = minerRoundVoting->GetVote(roundBlockIndex, update);
+      // LogPrint(BCLog::UPDATES, "Updates: MinerUpdateVoting::GetVote - retrieved vote for height %d: %s\n", roundBlockIndex->nHeight, roundVote.ToString());
+      vote += roundVote;
+   }
+   return vote;
+}
+
+VoteResult NodeUpdateVoting::GetVote(const CBlockIndex* blockIndex, const Update &update)
+{
+   // assert(height % update.roundSize == 0);
+   if (blockIndex->nHeight < update.StartHeight() + update.RoundSize() * update.VotingPeriod())
+   {
+      return VoteResult(0, update.RoundSize() * update.VotingPeriod()); // Assume everyone voted no for the entire period
+   }
+   // TODO: Cache lookup for Update Height
+
+   // Simply collect all of the round votes for a full average in the voting period:
+   VoteResult vote;
+   for (int lookback = 0; lookback < update.VotingPeriod(); ++lookback)
+   {
+      const CBlockIndex* roundBlockIndex = blockIndex->GetAncestor(blockIndex->nHeight - lookback * update.RoundSize());
+      if (roundBlockIndex == nullptr)
+         break; // Nothing more
+      VoteResult roundVote = nodeRoundVoting->GetVote(roundBlockIndex, update);
+      // LogPrint(BCLog::UPDATES, "Updates: NodeUpdateVoting::GetVote - retrieved vote for height %d: %s\n", roundBlockIndex->nHeight, roundVote.ToString());
+      vote += roundVote;
+   }
+   return vote;
+}
+
+
+UpdateManager::UpdateManager() :
+   minerRoundVoting(),
+   nodeRoundVoting(),
+   minerUpdateVoting(&minerRoundVoting),
+   nodeUpdateVoting(&nodeRoundVoting)
+{
+}
+
+UpdateManager::~UpdateManager() {};
+
+bool UpdateManager::Add(Update update)
+{
+   // TODO: Check for existence first
```

Comment:
```suggestion
  //Check for existence first and erase if exist
   auto it = updates.find(update.UpdateId());
   if (it != updates.end())
      updates.erase(it);
```
during start up 4 chains are initialized and the last chain initialized is the active chain, we only need update settings for the active chain

Reply:
Will fix - thanks.